// @jamesgmarks/utilities/dist/miscellaneous.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var InvalidParameterError_1 = require("./misc/types/InvalidParameterError");
exports.InvalidParameterError = InvalidParameterError_1.InvalidParameterError;
var boolVal_1 = require("./misc/boolVal");
exports.boolVal = boolVal_1.boolVal;
var def_1 = require("./misc/def");
exports.def = def_1.def;
exports.isDefined = def_1.isDefined;
exports.isDef = def_1.isDef;
var empty_1 = require("./misc/empty");
exports.empty = empty_1.empty;
var envToType_1 = require("./misc/envToType");
exports.envToType = envToType_1.envToType;
var floatVal_1 = require("./misc/floatVal");
exports.floatVal = floatVal_1.floatVal;
var intVal_1 = require("./misc/intVal");
exports.intVal = intVal_1.intVal;
var isBigInt_1 = require("./misc/isBigInt");
exports.isBigInt = isBigInt_1.isBigInt;
var isBoolean_1 = require("./misc/isBoolean");
exports.isBoolean = isBoolean_1.isBoolean;
var isBoolString_1 = require("./misc/isBoolString");
exports.isBoolString = isBoolString_1.isBoolString;
var isIntegerNumber_1 = require("./misc/isIntegerNumber");
exports.isIntegerNumber = isIntegerNumber_1.isIntegerNumber;
var isNull_1 = require("./misc/isNull");
exports.isNull = isNull_1.isNull;
var isNumber_1 = require("./misc/isNumber");
exports.isNumber = isNumber_1.isNumber;
var isNumeric_1 = require("./misc/isNumeric");
exports.isNumeric = isNumeric_1.isNumeric;
var isPrimitive_1 = require("./misc/isPrimitive");
exports.isPrimitive = isPrimitive_1.isPrimitive;
var isReference_1 = require("./misc/isReference");
exports.isReference = isReference_1.isReference;
var isString_1 = require("./misc/isString");
exports.isString = isString_1.isString;
var isSymbol_1 = require("./misc/isSymbol");
exports.isSymbol = isSymbol_1.isSymbol;
var isValidJSON_1 = require("./misc/isValidJSON");
exports.isValidJSON = isValidJSON_1.isValidJSON;
var isValidNumber_1 = require("./misc/isValidNumber");
exports.isValidNumber = isValidNumber_1.isValidNumber;
var notEmpty_1 = require("./misc/notEmpty");
exports.notEmpty = notEmpty_1.notEmpty;
var notNull_1 = require("./misc/notNull");
exports.notNull = notNull_1.notNull;
var sameType_1 = require("./misc/sameType");
exports.sameType = sameType_1.sameType;
var stringToBool_1 = require("./misc/stringToBool");
exports.stringToBool = stringToBool_1.stringToBool;
var strVal_1 = require("./misc/strVal");
exports.strVal = strVal_1.strVal;
var undef_1 = require("./misc/undef");
exports.undef = undef_1.undef;
exports.notDef = undef_1.notDef;
exports.isUndefined = undef_1.isUndefined;
var typeCast_1 = require("./misc/typeCast");
exports.typeCast = typeCast_1.typeCast;
